### DISCUSSION TITLE
How can I trigger a retry if  the consumption of some message fails for a retryable reason?

### DISCUSSION THREAD

**Discussion by daniel-bray-sonalake** [rabbitmq/rabbitmq-stream-java-client#106] [Q&A] [answered]
Hi all, thanks in advance for any help on this one. 

We have a reasonably likely race condition in our own application where

* We partially create something, and send out a message to some external service to validate some part of it.
* We don't want to move the message sending work outside the transaction that creates the entity, because we're implementing a transactional outbox.
* We sometimes get the response message for that validating message before the transaction that created the related entity is closed

So I was wondering: is it possible to have the stream consumer's `messageHandler` throw some class of error, or in some other way tell the consumer to retry a given message after some short time?

I tried this but the exception is just swallowed in `com.rabbitmq.stream.impl.ServerFrameHandler.BaseFrameHandler`

```
return streamEnvironment.consumerBuilder()
        .name(name)
        .offset(OffsetSpecification.first())
        .stream(stream)
        .autoTrackingStrategy().builder()
        .messageHandler((context, message) -> {
            var entity = findFor(message);
            if (null == entity) {
                throw new EntityNotFoundException("No entity for xxx, try again soon");
            } else {
                doStuff(message);
            }
        })
        .build();
```        

Thanks

**Comment by acogoluegnes** [comment, marked as answer]
Sorry for the late answer, somehow I'm not notified of new discussion entries :-(

I don't think retry is something the client library should handle. Retry can have complicated semantics, so I'd let this responsibility to the application, with its own utilities or retry frameworks [1] [2].

[1] https://github.com/spring-projects/spring-retry
[2] https://resilience4j.readme.io/docs/retry
